Kudu.Services/Diagnostics/ProcessController.cs

```diff
@@ -222,6 +236,199 @@ public HttpResponseMessage MiniDump(int id, int dumpType = 0, string format = nu
             }
         }
 
+        [HttpPost]
+        public async Task<HttpResponseMessage> StartProfileAsync(int id)
+        {
+            using (_tracer.Step("ProcessController.StartProfileAsync"))
+            {
+                // Check if the process Id exists in the sandbox. If it doesn't, this methid returns a 404 and we are done.
+                var process = GetProcessById(id);
+
+                // Check if the profiling is already running for the given process. If it does, then just return with 200.
+                if (_profilingList.ContainsKey(process.Id))
+                {
+                    return Request.CreateResponse(HttpStatusCode.OK);
+                }
+
+                // Profiling service supports up to 255 concurrent profiling sessions.
+                if (_profilingList.Count >= 255)
+                {
+                    _tracer.TraceError("Too many profiling sessions are currently running.");
+                    return Request.CreateErrorResponse(HttpStatusCode.ServiceUnavailable, "Too many profiling sessions are currently running. Please try again later.");
+                }
+
+                int profilingSessionId = GetNextProfilingSessionId();
+
+                string processName = System.Environment.ExpandEnvironmentVariables("%SystemDrive%\\msvsmon\\profiler\\VSStandardCollector.Dev14.exe");
+                string arguments = System.Environment.ExpandEnvironmentVariables(string.Format("start {0} /attach:{1} /loadAgent:4EA90761-2248-496C-B854-3C0399A591A4;DiagnosticsHub.CpuAgent.dll  /scratchLocation:%LOCAL_EXPANDED%\\Temp", profilingSessionId, id));
+
+                var profileProcessResponse = await HandleProfilingProcessAsync(processName, arguments);
+
+                if (!profileProcessResponse.IsSuccessStatusCode)
+                {
+                    return profileProcessResponse;
+                }
+
+                // This only fails in rare cases when two or more threads start profiling sessions for same process as same time
+                // In thast case we try to stop the other profile session except first one and return 500
+                // The profiling service already handles the case where two different processes try to start profiling with same session Ids. The sessions will fail except first one.
+                if(!_profilingList.TryAdd(process.Id, new ProfileInfo(profilingSessionId)))
+                {
+                    await StopProfileAsync(id, false);
+                    _tracer.TraceError("Failed to start profiling session.");
+                    return Request.CreateErrorResponse(HttpStatusCode.InternalServerError, "Failed to start profiling session because there were concurrent requests to start profiling sessions. Please try again later.");
```

Comment:
Wondering why this is an error case, and instead isn't a 200 response as you return higher up in the method in the case of conflicts?